### PR TITLE
Support SPN attribute for signals

### PR DESCRIPTION
**Pull request by approximatenumber**
Added SPN attribute to signal attributes.
Added SPN support to tests. But currently tests are broken in `master`, only python-tests are passed successfully.

### CODE REVIEW

File: cantools/database/can/signal.py [hunk 1]
```diff
@@ -387,6 +389,19 @@ def multiplexer_signal(self):
     def multiplexer_signal(self, value):
         self._multiplexer_signal = value
 
+    @property
+    def spn(self):
+        """The SPN (Suspect Parameter Number) value if the signal
+        has this attribute, ``None`` otherwise.
```
Comment (eerimoq) [inline cantools/database/can/signal.py:395]: You should mention J1939.

Comment (approximatenumber) [reply cantools/database/can/signal.py:395]: You may that I need to validate that DBC file is based on J1939 before SPN parsing?

Comment (eerimoq) [reply cantools/database/can/signal.py:395]: Nothing like that. Just Mention J1939 in the docstring. SPN is part of the J1939 specification.

Comment (approximatenumber) [reply cantools/database/can/signal.py:395]: Got it, mentioned.

**Comment by eerimoq** [comment]
Just add a test case (and/or extend an existing test case) and I'll merge.

**Comment by approximatenumber** [comment]
> Just add a test case (and/or extend an existing test case) and I'll merge.

Added SPN attribute test to existing `test_attributes`
Added SPN attribute to `attributes.dbc`

**Comment by eerimoq** [comment]
I'm confused. You add the new attribute ``spn`` to the Signal class, but the test does not test that at all. Please test the new code, nothing else. =)

**Comment by approximatenumber** [comment]
ok, should I leave attribute test and SPN values in DBC file?

**Comment by eerimoq** [comment]
I suggest removing that. Let's just add a test of the new code now. Add SPN to a signal in `tests/files/dbc/j1939.dbc` and check it in the test `test_j1939`.

**Comment by approximatenumber** [comment]
> I suggest removing that. Let's just add a test of the new code now. Add SPN to a signal in `tests/files/dbc/j1939.dbc` and check it in the test `test_j1939`.

Yep, done

**Comment by geo4711** [comment]
Hi!
Sorry for barging in ...
I've seen that the SPN is read from the attribute list. But I haven't found a piece of code that writes back the SPN value to an attribute.
Of course, this only matters if you change the member variable spn via the setter methode ...

Did I miss anything about that? Perhaps you could double check that.
Thanks :-)

**Comment by eerimoq** [comment]
I'm open to adding dump functionality later. Just reading is a good start.

File: tests/test_database.py [hunk 2]
```diff
@@ -3880,6 +3881,20 @@ def test_j1939_dbc(self):
...
+        # test SPN signal attribute
+        attribute = attributes['SPN']
+        self.assertEqual(attribute.name, 'SPN')
+        self.assertEqual(attribute.value, 500)
+        self.assertEqual(attribute.definition,
+                         db.dbc.attribute_definitions['SPN'])
+        self.assertEqual(attribute.definition.kind, 'SG_')
+        self.assertEqual(attribute.definition.type_name, 'INT')
+        self.assertEqual(attribute.definition.minimum, 0)
+        self.assertEqual(attribute.definition.maximum, 524287)
```
Comment (eerimoq) [inline tests/test_database.py:3895]: Remove all the above.

**Comment by approximatenumber** [comment]
> I can't find any test verifying that the SPN loaded from the dbc-file can be accessed using the spn attribute on the signal class. That is the most relevant test you can add, . Please add it.

Sorry for misunderstanding. Please, check. Just added a simple accessability test.

**Comment by RangelReale** [comment]
Shouldn't the SPN attribute use the default value if available? Currently it doesn't take it in account.

In the test file `dbc/j1939.dbc`, we have

```
BA_DEF_DEF_  "SPN" 0;
```

but on the test we check

```python
        signal = db.messages[1].signals[0]
        self.assertEqual(signal.spn, None)
```

shouldn't it be 0?

**Pull request merged**